src/const_eval.md

```diff
@@ -81,11 +81,81 @@ r[const-eval.const-expr.builtin-arith-logic]
   operators used on integer and floating point types, `bool`, and `char`.
 
 r[const-eval.const-expr.borrows]
-* All forms of [borrow]s, including raw borrows, with one limitation:
-  mutable borrows and shared borrows to values with interior mutability
-  are only allowed to refer to *transient* places or to *static* places. A place is *transient*
-  if its lifetime is strictly contained inside the current [const context].
-  A place is *static* if it is a `static` item or a [promoted expression].
+* All forms of [borrow]s, including raw borrows, with one limitation: mutable borrows and shared borrows of expressions whose temporary scope would be extended (see [temporary lifetime extension]) to the end of the program are not allowed when those borrows refer to values with interior mutability.
+
+  ```rust,compile_fail,E0492
+  # use core::sync::atomic::AtomicU8;
+  const C: &AtomicU8 = &AtomicU8::new(0); // ERROR not allowed
+  ```
+
+  ```rust,compile_fail,E0764
+  # use core::sync::atomic::AtomicU8;
+  const C: &AtomicU8 = &mut AtomicU8::new(0); // ERROR not allowed
+  ```
+
+  ```rust,compile_fail,E0492
+  # use core::sync::atomic::AtomicU8;
+  let _: &'static _ = const { &AtomicU8::new(0) }; // ERROR not allowed
+  ```
+
+  ```rust
+  # use core::sync::atomic::AtomicU8;
+  const C: () = { _ = &AtomicU8::new(0); }; // OK
+  ```
+
+  ```rust
+  # use core::sync::atomic::AtomicU8;
+  static S: AtomicU8 = AtomicU8::new(0);
+  const C: &AtomicU8 = &S; // OK
+  ```
+
+  In other words, they are only allowed to refer to *transient* places, to *indirect* places, or to *static* places. A place is *transient* if it is based on a local variable whose lifetime is strictly contained inside the current [const context].
+
+  ```rust
+  const C: () = {
+      let mut x = 0;
+      let _ = &mut x; // Reference to a transient place.
+  };
+  ```
```

Comment:
When we say, @RalfJung, "local variable" above, are we meaning to exclude non-promoted non-lifetime-extended temporaries from the definition of transient?  Intuitively, those seem transient.

Reply:
Hm... there's some friction here since the analysis runs on MIR, where there's no distinction between explicitly `let`-bound variables and temporaries. In MIR, literally what we do is check that the place expression is of the form `$local.$projections` where the projections do not contain a `*`, and then we check that `local` is a local variable that has a `StorageDead` which indicates that it has not been lifetime-extended. (If the projections do contain a `*`, then this is an easy case, we can just allow it as this is an indirect place.)

In terms of the surface language, this means indeed non-lifetime-extended temporaries are transient. Promotion is irrelevant, as we recently learned.

Reply:
> How does the current impl think about this, as transient or as static?

It thinks of this as transient. It doesn't see the lifetimes, and the MIR at this point looks like
```rust
const C: () = {
    let mut arr = [];
    let _: &mut [u8] = &mut arr;
};
```
> Still, I'm curious how that interacts with that StorageDead check you mention, as presumably that wouldn't be there due to the promotion.

It is there. Promotion later removes it.

Reply:
Ah, removed later.  Got it, thanks.

Reply:
For some context: MIR building knows nothing about promotion. It compiles that code like it would compile `let x: &mut [u8] = &mut [13];`. Promotion is a MIR pass that runs later (between const check and borrow checking, apparently) that does some heavy patching of the MIR to remove a local and its initializing expression and replace it by references to a new anonymous const.